src/ruby/puppet-server-lib/puppet/server/puppet_config.rb

```diff
@@ -21,12 +21,15 @@ def self.initialize_puppet(puppet_config)
     )
     Puppet[:trace] = true
 
-    # (SERVER-410) Cache features in puppetserver for performance.  Avoiding
-    # the cache is intended for agents to reload features mid-catalog-run.
-    Puppet[:always_cache_features] = true
-
+    # This conditional can be replaced with just setting always_retry_plugins
+    # once puppetserver depends on a version of puppet that has that setting
+    # available.
     if Puppet.settings.setting('always_retry_plugins')
       Puppet[:always_retry_plugins] = false
+    else
+      # (SERVER-410) Cache features in puppetserver for performance.  Avoiding
+      # the cache is intended for agents to reload features mid-catalog-run.
+      Puppet[:always_cache_features] = true
```

Comment:
Depends if the server and agent releases are coordinated/coupled or not. Leaving it this way would allow us to ship a Y even if puppet 4.5 hasn't shipped.